kernel: Add new platform EN75xx and device SmartFiber XP8421-B

cjdelisle opened the pull request:
EcoNet EN75xx is a big endian MIPS platform used in XPON (fiber), DSL, and SIM (3g/4g) applications. Complete GPL vender SDKs exist for this platform, but are based on Linux 2.6.

Most of this PR is a backport of patches which are already accepted in upstream. What's added is SPI and Bad Block Table that are based on a MediaTek BMT framework that is not upstream.

The SmartFiber XP8421-B is a fiber modem which is available for $20 online and has 512MB of memory, 256MB of SPI NAND flash and 2 USB 2.0 ports in addition to ethernet, wifi and XPON.

Many of the drivers for this SoC bear little resemblence to anything currently supported in OpenWRT or Linux upstream, so there is no realistic timeline for reaching a feature complete port. The aim of this patch is to provide the foundation (INTC, timer, SPI, NAND, image format) to be able to flash and boot to a shell.

I'm not getting paid for this, I'm just a hobbyist, and this is my first contribution. So any advice and guidance to help get this landed will be most appreciated.

The XP8421-B, and apparently many other devices of this platform, use a dual-image layout. I have chosen to reuse this to support dual-boot between OpenWRT and the factory firmware. Certain design decisions were made with the goal of not overwriting data that is used by the factory OS.

Flashing instructions (from bootloader):
1. Build and then locate the squashfs-tclinux.trx image file
2. Get the length of that file in hex: `printf '%X\n' "$(stat -c%s the-file-squashfs-tclinux.trx)"`
3. Connect to device with xmodem capability, e.g. `picocom --send-cmd lsx -vv -b 115200 /dev/ttyUSB0`
4. Switch device on and press a key within 3 seconds
5. Enter bootloader username and password: telecomadmin nE7jA%5m
6. Type: `xmdm 80020000 <file length hex>`
7. Quickly start xmodem and send the file, in picocom that is `ctrl+a ctrl+s <paste-the-file-name> enter` If the transfer fails to start, wait 30 seconds to a minute for the bootloader prompt to return and then try the command again.
8. Once the transfer has completed successfully, type the following `flash 80000 80020000 <file length hex>`
9. Type `go` or simply restart the device to boot into OpenWRT

Djfe commented on target/linux/econet/files/drivers/mtd/nand/en75_bmt.c:
```diff
@@ -0,0 +1,1439 @@
+ * effect, data loss. Furthermore, bad blocks do not politely identify
+ * themselves, generally they error out when we try to read them. Still, we
+ * make the best effort we can by tagging worn blocks with 0x55 rather than
+ * 0x00 which is used to tag factory blocks. Vender firmware does not do this,
```
*Vendor

cjdelisle replied:
Fixed all in push -f, thanks !

andrewjlamarche commented on target/linux/econet/Makefile:
```diff
@@ -0,0 +1,22 @@
+# include the profiles
+include $(INCLUDE_DIR)/target.mk
+
+$(eval $(call BuildTarget))
```
missing newline

andrewjlamarche commented on target/linux/econet/base-files/sbin/en75_chboot:
```diff
@@ -0,0 +1,112 @@
+        exit 1
+    fi
+}
+main "$@"
```
missing newline

cjdelisle replied:
Fixed, thanks.

andrewjlamarche commented on target/linux/econet/dts/en751221_tplink_archer-vr1200v-v2.dts:
```diff
@@ -0,0 +1,69 @@
+        #size-cells = <1>;
+
+        partition@0 {
+            label = "bootloader";
+            reg = <0x0 0x00080000>;
+            read-only;
+        };
+		partition@80000 {
+			label = "misc";
+			reg = <0x00080000 0x140000>;
+		};
+        partition@1c0200 {
+            label = "kernel";
+            reg = <0x001c0000 0x400000>;
+        };
+		partition@5c0000 {
+			label = "rootfs";
+			reg = <0x005c0000 0x1a40000>;
+			linux,rootfs;
+		};
+		partition@1c0000 {
+			label = "firmware";
+			reg = <0x001c0000 0x1e40000>;
+		};
+		partition@2000000 {
+			label = "firmware_factory";
+			reg = <0x2000000 0x1e40000>;
+		};
+		partition@3fe0000 {
+			label = "reserve";
+			reg = <0x3fe0000 0x20000>;
+		};
+		partition@4000000 {
+			label = "openwrt_ubi";
+			/* From the factory this is unallocated space, so it's ours for the taking.
+			 * We have up to 0x35e0000 of space, but we reserve 10 eraseblocks (1.25MB)
+			 * in case the BMT steals them from us.
+			 */
+			reg = <0x4000000 0x34a0000>;
+		};
```
There's some funky indents here. Could you address that?

andrewjlamarche commented on target/linux/econet/dts/en751221_generic.dts:
Is there a point of having a generic dts? There are a lot of unknowns in this.

cjdelisle replied:
I use this for getting Linux onto a device that doesn't (yet) have it's own dts. The process is you build a ramfs kernel and send it to the bootloader with xmodem (these bootloaders don't support much, but they have xmodem), then you jump to it.

Accessing memory that isn't there is no good, and the last thing you want is for it to try to load the BBT and potentially trash your disk (Ask Me How I Know), so this image will just prove that the device can work, and give you a way to dump the disk.

We could omit it, but it will be in the instructions for adding a new econet device.

andrewjlamarche commented:
There is a lot of good work here, nice job! 

I would recommend splitting this into a couple PRs; one adding platform support and one for supporting each of the devices. This way it is cleaner to follow the commit log.

andrewjlamarche commented:
Hi! I think it’s no problem to have everything in the same PR, but splitting into separate commits is preferred so each one has a very specific change and should problems arise later on, it is much easier to bisect/debug. 

Also, historically, each device support has its own commit with flashing and recovery instructions.

cjdelisle commented:
Ok that sounds good. What you do say to adding base platform + Generic in 1, then adding the Smartfiber in 2 then the TpLink in 3? That way every commit will contain code which is capable of building *something*...

cjdelisle commented:
@andrewjlamarche updated per your recommendation. 3 commits, one for the platform only, then Smartfiber, then TPLink. Generic is removed from this iteration.

cjdelisle commented:
I definitely tested when I developed on them, but I'll do a final test of precisely what is in this version of the PR.

As far as ETH, PCI, USB, these are going to take some time. The PCI and USB are similar to ramips but not identical, and the ETH subsystem is a totally different beast from anything found in open source (except Airoha which has something fairly similar). Then the XPON, DSL, and VoIP are totally proprietary and have no timeline for support at all.

The point of making this PR now is to have a common workplace where EcoNet enthusiasts can build out the rest of the driver ecosystem in one common place. There have been a number of attempts to port recent OpenWRT to EcoNet in the past, but they relied on OpenWRT forks and patched kernels which inevitably went out of date and lost momentum. This is just a large project, so without baby steps I don't see any realistic path to completion.

A more detailed description of the boards can be found here and here:
* https://openwrt.org/inbox/toh/evaluation_boards/unbranded_boards/smartfiber_xp8421-b#experimental_firmware
* https://openwrt.org/inbox/toh/tp-link/archer_vr1200v#experimental_firmware

cjdelisle commented on target/linux/econet/patches-6.12/001-dt-bindings-vendor-prefixes-Add-EcoNet.patch:
```diff
@@ -0,0 +1,28 @@
+From 53d6a6489776e5abe34182f4425f364d127403fe Mon Sep 17 00:00:00 2001
```
Oh yes, sure thing. I just took the patches as I mailed them in. I'll get that switched up.

abajk commented:
> @Rondom I've just updated the patches to the ones taken from the linux git and also rebased.

Consider adding versions to backported patches, as other targets do. You should also refresh the patches.

abajk commented on target/linux/econet/patches-6.12/300-spi-Add-support-for-the-Airoha-EN7523-SoC-SPI-contro.patch:
```diff
@@ -0,0 +1,341 @@
+--- a/drivers/spi/Kconfig
```
There are also a few checkpatch warnings:
~~~
aleksander@laptop-olek:~/workspace/linux$ scripts/checkpatch.pl --strict ~/workspace/econet/target/linux/econet/patches-6.12/300-spi-Add-support-for-the-Airoha-EN7523-SoC-SPI-contro.patch 
...
WARNING: please write a help paragraph that fully describes the config symbol with at least 4 lines
#7: FILE: drivers/spi/Kconfig:373:
+config SPI_AIROHA_EN7523
+	bool "Airoha EN7523 SPI controller support"
+	depends on ARCH_AIROHA
+	help
+	  This enables SPI controller support for the Airoha EN7523 SoC.
+

CHECK: Please don't use multiple blank lines
#36: FILE: drivers/spi/spi-en7523.c:8:
+
+

CHECK: Macro argument 'x' may be better as '(x)' to avoid precedence issues
#87: FILE: drivers/spi/spi-en7523.c:59:
+#define REG(x)			(iobase + x)

CHECK: Please don't use multiple blank lines
#89: FILE: drivers/spi/spi-en7523.c:61:
+
+

CHECK: Please don't use multiple blank lines
#92: FILE: drivers/spi/spi-en7523.c:64:
+
+

WARNING: Consider removing the code enclosed by this #if 0 and its #endif
#169: FILE: drivers/spi/spi-en7523.c:141:
+#if 0

CHECK: Prefer kernel type 'u8' over 'uint8_t'
#199: FILE: drivers/spi/spi-en7523.c:171:
+	uint8_t *buf = xfer->rx_buf;

CHECK: Prefer kernel type 'u8' over 'uint8_t'
#222: FILE: drivers/spi/spi-en7523.c:194:
+	const uint8_t *buf = xfer->tx_buf;

CHECK: Logical continuations should be on the previous line
#272: FILE: drivers/spi/spi-en7523.c:244:
+			if (!list_is_last(&xfer->transfer_list, &msg->transfers)
+			    && list_next_entry(xfer, transfer_list)->rx_buf != NULL)

CHECK: Comparison to NULL could be written "rx_buf"
#272: FILE: drivers/spi/spi-en7523.c:244:
+			    && list_next_entry(xfer, transfer_list)->rx_buf != NULL)

WARNING: DT compatible string "airoha,en7523-spi" appears un-documented -- check ./Documentation/devicetree/bindings/
#324: FILE: drivers/spi/spi-en7523.c:296:
+	{ .compatible = "airoha,en7523-spi" },

WARNING: Prefer "GPL" over "GPL v2" - see commit bf7fbeeae6db ("module: Cure the MODULE_LICENSE "GPL" vs. "GPL v2" bogosity")
#339: FILE: drivers/spi/spi-en7523.c:311:
+MODULE_LICENSE("GPL v2");
...
~~~

cjdelisle replied:
Given this is code copied from elsewhere in OpenWRT, please let me know if re-working it as my own is something that you want me to do.

Rondom replied:
I stand corrected then and I agree if DMA is indeed not supported that it will be easier with a separate driver. I myself am not familiar with the differences between Econet MIPS chips.

abajk commented:
Despite the comments, this PR looks promising.

abajk commented on target/linux/econet/image/Makefile:
```diff
@@ -0,0 +1,51 @@
+endef
+TARGET_DEVICES += tplink_archer-vr1200v-v2
+
+define Device/smartfiber_xp8421-b
```
Sorting. Smartfiber should be before tp-link :)

cjdelisle replied:
Will fix.

cjdelisle commented on target/linux/econet/patches-6.12/301-spi-Airoha-adapt-to-support-en75-mips.patch:
```diff
@@ -0,0 +1,44 @@
+--- a/drivers/spi/Kconfig
```
Ok will fix.

cjdelisle replied:
Fixed (see comment below)

pstlr78 commented:
Is it so difficult (or scary ?) for someone to review the changes and push this PR though?
This is a great piece of work from @cjdelisle and many enthusiasts are waiting for years now for this to happen. This will open the doors for supporting dozens of other devices of the same or very similar Econet family.

aparcar commented:
@cjdelisle could you please provide a link where to buy such device?

cjdelisle commented:
@aparcar You can buy them from the factory here: https://www.ismartfiber.com/index.php/index/show?id=135&catname=XPONONT you have to email them and wire the funds, but they're nice folks, they'll ship you a single unit.

cjdelisle commented:
@aparcar Thanks! I think I set it as source-only, so unless I'm mistaken, it shouldn't be stressing the build bots.

abajk commented on target/linux/econet/dts/en751221.dtsi:
```diff
@@ -0,0 +1,82 @@
+
+		cpu@0 {
+			device_type = "cpu";
+			compatible = "mips,mips24KEc";
```
I believe it is mips 34Kc.

abajk replied:
MT7620 is a MIPS 24Kc core. MIPS 34Kc is used in Realtek RTL839x, Lantiq xRX200, etc. This string currently isn't used in the kernel, so it doesn't affect anything.

Djfe replied:
good decision

abajk commented:
LGTM

pstlr78 commented:
Can someone review and merge this PR to snapshot/main/master? Next week it is 3 months since it was first opened.

cjdelisle committed:
kernel: Add new platform EcoNet MIPS

EcoNet EN75xx is a big endian MIPS platform used in XPON (fiber),
DSL, and SIM (3g/4g) applications. Complete GPL vender SDKs exist
for this platform, but are based on Linux 2.6.

The bulk of this submission has already been accepted upstream:
https://patchwork.kernel.org/project/linux-mips/list/?series=960479&state=*

This platform uses a bootloader that is derived from old TrendChip
code. This bootloader implements a frustratingly complex Bad Block
Table which is implemented here in en75_bmt.c

This BMT is not upstreamed because it depends on mtk_bmt framework
which likewise is not upstreamed.

This BMT system rewrites block indexes in flash and if the bootloader
considers it to be corrupted, it will attempt to automatically rebuild
on boot. So without implementing the algorithm, you can't safely use
the disk at all.

Signed-off-by: Caleb James DeLisle <cjd@cjdns.fr>
Link: https://github.com/openwrt/openwrt/pull/19021
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

cjdelisle committed:
econet: Add new target SmartFiber XP8421-B

The SmartFiber XP8421-B is a fiber modem which is available for $20 online
and has 512MB of memory, 256MB of SPI NAND flash and 2 USB 2.0 ports in
addition to ethernet, wifi and XPON.

Because EcoNet is not currently producing evaluation boards, the XP8421-B
stands in as a convenient, low cost, off-the-shelf, representitive example
of the capabilities of the EN751221 econet processor. This is also the
example board that is included in the upstream Linux patchset.

The XP8421-B, and apparently many other devices of this platform, use a
dual-image layout. I have chosen to reuse this to support dual-boot between
OpenWRT and the factory firmware. Certain design decisions were made with
the goal of not overwriting data that is used by the factory OS.

This commit also introduces a utility for switching between OS_A and OS_B
which are used for OpenWRT and Factory OS respectively.

Flashing instructions (from bootloader):

Build and then locate the squashfs-tclinux.trx image file
Get the length of that file in hex: printf '%X\n' "$(stat -c%s the-file-squashfs-tclinux.trx)"
Connect to device with xmodem capability, e.g. picocom --send-cmd lsx -vv -b 115200 /dev/ttyUSB0
Switch device on and press a key within 3 seconds
Enter bootloader username and password: telecomadmin nE7jA%5m
Type: xmdm 80020000 <file length hex>
Quickly start xmodem and send the file, in picocom that is ctrl+a ctrl+s <paste-the-file-name> enter If the transfer fails to start, wait 30 seconds to a
minute for the bootloader prompt to return and then try the command again.
Once the transfer has completed successfully, type the following flash 80000 80020000 <file length hex>
Type go or simply restart the device to boot into OpenWRT

Signed-off-by: Caleb James DeLisle <cjd@cjdns.fr>
Link: https://github.com/openwrt/openwrt/pull/19021
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

cjdelisle committed:
econet: Add new target TP-Link Archer VR1200v (v2)

The TP-Link Archer VR1200v (v2) is a low end DSL modem based on the
EcoNet EN751221 processor platform.

While it does have an unlocked bootloader, the factory upgrade feature
requires a cryptographic signature so flashing from the web UI is not
feasible.

The Archer VR1200v (v2) uses a dual-image layout. I have chosen to reuse
this to support dual-boot between OpenWRT and the factory firmware.

Flashing instructions (from bootloader):

Build and then locate the squashfs-sysupgrade.bin image file
Get the length of that file in hex: printf '%X\n' "$(stat -c%s the-file-squashfs-sysupgrade.bin)"
Connect to device with xmodem capability, e.g. picocom --send-cmd lsx -vv -b 115200 /dev/ttyUSB0
Switch device on and press a key within 3 seconds, you should get to a `bldr>` prompt
Type: xmdm 80020000 <file length hex>
Quickly start xmodem and send the file, in picocom that is ctrl+a ctrl+s <paste-the-file-name> enter If the transfer fails to start, wait 30 seconds to a minute for the bootloader prompt to return and then try the command again.
Once the transfer has completed successfully, type the following flash 80000 80020000 <file length hex>
Type `re` or simply restart the device to boot into OpenWRT

Signed-off-by: Caleb James DeLisle <cjd@cjdns.fr>
Link: https://github.com/openwrt/openwrt/pull/19021
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

Djfe commented:
> Thanks. I'm okay merging this but would like to get at least one more core de eloped to approach adding a new target. It's a considerable workload for our build infrastructure

@robimarko could you be the one to also take a look at this PR please so @cjdelisle can continue working on the next steps of this new target? :)
@apacar Right now it is being merged as source-only so there should be no additional load on the build infrastructure in the next months. But of course the expectation is to eventually add this as a working target that is being built by the build bots, too. 😅

fzdy-zz commented:
@cjdelisle 
Yes, a very great plan.
I found this place after reading it.

an7581 is used because it’s relatively simple.
Currently, xgpon && xepon use an7581, which has a lot of affordable devices. More importantly, you can directly use Airoha’s SDK to operate the PON.

```
https://dev.iopsys.eu/feed/targets/-/tree/devel/airoha
```

Wishing the plan goes smoothly.
